doc/contributing.rst

```diff
@@ -1,221 +1,808 @@
+    ``:func:`mne.set_config``` will work but ``:func:`mne.utils.set_config```
+    will not).
+
+
+Other style guidance
+~~~~~~~~~~~~~~~~~~~~
+
+- Use single quotes whenever possible.
+
+- Prefer :ref:`generators <tut-generators>` or
+  :term:`comprehensions <list comprehension>` over :func:`filter`, :func:`map`
+  and other functional idioms.
+
+- Use explicit functional constructors for builtin containers to improve
+  readability (e.g., :ref:`list() <func-list>`, :ref:`dict() <func-dict>`,
+  :ref:`set() <func-set>`).
+
+- Avoid nested functions or class methods if possible — use private functions
+  instead.
+
+- Avoid ``*args`` and ``**kwargs`` in function/method signatures.
+
+
+Code organization
+^^^^^^^^^^^^^^^^^
+
+Importing
+~~~~~~~~~
+
+Import modules in this order:
+
+1. Python built-in (``os``, ``copy``, ``functools``, etc)
+2. standard scientific (``numpy as np``, ``scipy.signal``, etc)
+3. others
+4. MNE-Python imports (e.g., ``from .pick import pick_types``)
+
+When importing from other parts of MNE-Python, use relative imports in the main
+codebase and absolute imports in the tutorials and how-to examples. Imports for
+``matplotlib`` and optional modules (``sklearn``, ``pandas``, etc.) should be
+nested (i.e., within a function or method, not at the top of a file).
+
+
+Return types
+~~~~~~~~~~~~
+
+Methods should modify inplace and return ``self``, functions should return
+copies (where applicable). Docstrings should always give an informative name
+for the return value, even if the function or method's return value is never
+stored under that name in the code.
+
+
+Vizualization
+~~~~~~~~~~~~~
+
+Visualization capabilities should be made available in both function and method
+forms. Add public visualization functions to the :mod:`mne.viz` submodule, and
+call those functions from the corresponding object methods. For example, the
+method :meth:`mne.Epochs.plot` internally calls the function
+:func:`mne.viz.plot_epochs`.
+
+All visualization functions must accept a boolean ``show`` parameter and return
+a :class:`~matplotlib.figure.Figure` handle.
+
+Visualization functions should default to the colormap ``RdBu_r`` for signed
+data with a meaningful middle (zero-point) and ``Reds`` otherwise. This applies
+to both visualization functions and tutorials/examples.
+
+
+.. _run_tests:
+
+Running the test suite
+^^^^^^^^^^^^^^^^^^^^^^
+
+Running the full test suite is as simple as running
 
 .. code-block:: console
 
-   $ PATTERN=plot_background_filtering.py make html_dev-pattern
+    $ make test
 
-Consult the `sphinx gallery documentation`_ for more details.
+from the ``mne-python`` root folder. Testing the entire module can be quite
+slow, however, so to run individual tests while working on a new feature, you
+can run, e.g.:
 
-Deprecating
------------
-If you need to deprecate a function or a class, use the ``@deprecated`` decorator::
+.. code-block:: console
+
+    $ pytest mne/tests/test_evoked.py:test_io_evoked -x --verbose
```

Comment:
can you say what the -x does? I never use it.

I would also mention that you can use the `--pdb` option to end up in the debugger on a crash.

Reply:
added explanation in a sidebar.  Also removed `-x` from the actual commands, since it doesn't really do anything when you're selecting a single test to run (as we show here).